### PR TITLE
Warn about objects with toString that get implicitly converted to strings

### PR DESCRIPTION
**Pull request by gaearon**
Complementary to #10385. This is not ideal but this encourages explicitness, and leaves an escape hatch for cases like Glamor rules or URI/i18n objects.

### CODE REVIEW

**Comment by gaearon** [comment]
Fixed failures.

File: src/renderers/dom/fiber/ReactDOMFiberComponent.js [hunk 1]
```diff
@@ -1013,28 +994,37 @@ var ReactDOMFiberComponent = {
...
-        } else if ((propertyInfo = DOMProperty.properties[propKey])) {
-          // $FlowFixMe - Should be inferred as not undefined.
-          extraAttributeNames.delete(propertyInfo.attributeName);
-          serverValue = DOMPropertyOperations.getValueForProperty(
-            domElement,
-            propKey,
-            nextProp,
-          );
+        } else if (DOMProperty.shouldSetAttribute(propKey, nextProp)) {
+          if ((propertyInfo = DOMProperty.getPropertyInfo(propKey))) {
```
Comment (leonardoanalista) [inline src/renderers/dom/fiber/ReactDOMFiberComponent.js]: variable assignment inside an `if` condition can be overlooked.

File: src/renderers/dom/shared/HTMLDOMPropertyConfig.js [hunk 2]
```diff
@@ -22,197 +22,79 @@ var HAS_OVERLOADED_BOOLEAN_VALUE =
...
-    useMap: 0,
+    // itemScope is for for Microdata support.
+    // See http://schema.org/docs/gs.html
+    itemScope: HAS_BOOLEAN_VALUE,
+    // These attributes must stay in the white-list because they have
+    // different attribute names (see DOMAttributeNames below)
+    acceptCharset: 0,
+    className: 0,
+    htmlFor: 0,
+    httpEquiv: 0,
```
Comment (leonardoanalista) [inline src/renderers/dom/shared/HTMLDOMPropertyConfig.js]: `allowTransparency` is expected to appear as lowercase on html according to: https://msdn.microsoft.com/en-us/library/ms533072(v=vs.85).aspx. 
Does the same happen to `httpEquiv` (which is expected to be `http-equiv` on html) ?

File: src/renderers/dom/shared/hooks/ReactDOMUnknownPropertyHook.js [hunk 3]
```diff
@@ -76,54 +57,100 @@ if (__DEV__) {
...
+      warning(
+        false,
+        'Directly setting property `innerHTML` is not permitted. ' +
+          'For more information, lookup documentation on `dangerouslySetInnerHTML`.',
+      );
+      warnedProperties[name] = true;
+      return true;
+    }
+
+    if (typeof value === 'number' && isNaN(value)) {
```
Comment (leonardoanalista) [inline src/renderers/dom/shared/hooks/ReactDOMUnknownPropertyHook.js]: there is a potential utils functions here `const isNumber = (value) => typeof value === 'number';` to be re-used. I am sure this is not the first, neither the last time you need to check for a number.

**Comment by gaearon** [comment]
Decided not to do it.

**Pull request closed without merging**